File: BM73_最长回文子串/javascript_solutions/Manacher.ts
export function getLongestPalindrome(A: string): number {
    const mp = new Array(A.length * 2 + 2);
    calculate(A, mp);

    let maxLen = 0;
    for (let i = 0; i < mp.length; i++) {
        maxLen = Math.max(maxLen, mp[i] - 1);
    }
    return maxLen;
}

function calculate(A, mp){
    let ms = '$#';
    for (let i = 0; i < A.length; i++) {
        ms += A.charAt(i);
        ms += "#";
    }

    let rightSide = 0;
    let index = 0;
    
    for (let i = 0; i < ms.length; i++) {
        mp[i] = rightSide > i ? Math.min(mp[2*index - i], rightSide - i) : 1;

        while(i - mp[i] >= 0 && i + mp[i] < ms.length && ms[i - mp[i]] == ms[i + mp[i]]){
            mp[i] += 1;
        }
        
        if(i + mp[i] > rightSide){
            rightSide = i + mp[i];
            index = i;
        }
    }
}

File: BM73_最长回文子串/java_solutions/Manacher.java
package BM73_最长回文子串.java_solutions;

public class Manacher {
    public int getLongestPalindrome (String A) {
        int[] mp = new int[2*A.length() + 2];

        calculate(A, mp);

        int maxLen = 0;
        for (int i = 0; i < mp.length; i++) {
            maxLen = Math.max(maxLen, mp[i] -1);
        }
        return maxLen;
    }

    private void calculate(String A, int[] mp){
        String ms  = "$#";
        for (int i = 0; i < A.length(); i++) {
            ms += A.charAt(i);
            ms += "#";
        }

        int rightSide = 0;
        int maxIndex = 0;

        for (int i = 0; i < ms.length(); i++) {
            mp[i] = rightSide > i ? Math.min(mp[2*maxIndex - i], rightSide - i) : 1;
            while(
                i - mp[i] >= 0 && 
                i + mp[i] < ms.length() && 
                ms.charAt(i - mp[i]) == ms.charAt(i + mp[i])
            ){
                mp[i] += 1;
            }

            if(i + mp[i] > rightSide){
                rightSide = i + mp[i];
                maxIndex = i;
            }
        }
    }
}
